### PR TITLE
Add repository_ctx.patch API

### PR DESCRIPTION
**Pull request by meteorcloudy**
This API is guarded under `--incompatible_use_native_patch`

This PR is based on https://github.com/bazelbuild/bazel/pull/8773, please only review commits from "Add repository_ctx.patch API"

Working towards: #8316

### CODE REVIEW

File: src/test/java/com/google/devtools/build/lib/bazel/repository/skylark/SkylarkRepositoryContextTest.java [hunk 1]
```diff
@@ -278,6 +281,44 @@ public void testRead() throws Exception {
...
+        "+++ foo\n" +
+        "@@ -1,1 +1,2 @@\n" +
+        " line one\n" +
+        "+line two\n", false, true, null);
+    context.patch(patchFile, 0, null);
+    testOutputFile(foo.getPath(), String.format("line one%nline two%n"));
+  }
+
+  @Test
+  public void testPatchFailed() throws Exception {
```
Comment (irengrig) [inline src/test/java/com/google/devtools/build/lib/bazel/repository/skylark/SkylarkRepositoryContextTest.java]: Great! Please add two more tests:  1,4 (3 is probably similar to 2)

https://github.com/bazelbuild/bazel/pull/8555/files/dc0dccc8bbc3e676695fc114aa20f0e921bca2ba#r299372582

File: src/main/java/com/google/devtools/build/lib/skylarkbuildapi/repository/SkylarkRepositoryContextApi.java [hunk 2]
```diff
@@ -294,6 +294,35 @@ public SkylarkExecutionResultApi execute(
...
+      parameters = {
+          @Param(
+              name = "patch_file",
+              allowedTypes = {
+                  @ParamType(type = String.class),
+                  @ParamType(type = Label.class),
+                  @ParamType(type = RepositoryPathApi.class)
+              },
+              doc =
+                  "Path of the patch file to apply."),
```
Comment (irengrig) [inline src/main/java/com/google/devtools/build/lib/skylarkbuildapi/repository/SkylarkRepositoryContextApi.java]: Please document similar to https://github.com/bazelbuild/bazel/blob/dc0dccc8bbc3e676695fc114aa20f0e921bca2ba/src/main/java/com/google/devtools/build/lib/skylarkbuildapi/repository/SkylarkRepositoryContextApi.java#L60

(path can be a string, path, or label, and can be relative)

Comment (meteorcloudy) [reply src/main/java/com/google/devtools/build/lib/skylarkbuildapi/repository/SkylarkRepositoryContextApi.java]: Done

File: src/test/java/com/google/devtools/build/lib/bazel/repository/skylark/SkylarkRepositoryContextTest.java (hunk 1 above)
Comment (meteorcloudy) [reply src/test/java/com/google/devtools/build/lib/bazel/repository/skylark/SkylarkRepositoryContextTest.java]: Done

**Pull request closed without merging**